[k8s] Kubernetes support

romilbhardwaj opened the pull request:
Adds Kubernetes support to SkyPilot. Still largely a WIP, expect rough edges and please provide feedback :)

To try out, create a Kubernetes cluster (either locally with Kind or some hosted service like GKE, instructions [here](https://github.com/skypilot-org/skypilot/blob/k8s_cloud/tests/kubernetes/README.md#running-a-local-development-cluster)), make sure ports 30000-32767 are accessible, and place the kubeconfig in `~/.kube/config` and run `sky check`.

Current status, TODOs, and roadmap can be found in this [doc](https://docs.google.com/document/d/1IrrkbtkX74gx1VHlPLjB9c_CTpuE4VZiB-QCXPehzqY/edit#bookmark=id.jh816bcfn0is).

Also adds two new CLI commands:
* `sky local up` create a local kubernetes cluster to run tasks locally. Requires docker and kind installed.
* `sky local down` tears down the local cluster.

**What's not supported in this PR:**
* Multi-node
* GPU support
* Supporting multiple Kubernetes clusters
* Docs (both for admins and users)
* Stopping clusters
* ... 

```
(base) ➜  ~ SKYPILOT_DEBUG=0 sky launch --cpus 8+
== Optimizer ==
Target: minimizing cost
Estimated cost: $0.0 / hour

Considered resources (1 node):
---------------------------------------------------------------------------------------------------
 CLOUD        INSTANCE          vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN
---------------------------------------------------------------------------------------------------
 Kubernetes   8vCPU-8GB         8       8         -              kubernetes    0.00          ✔
 AWS          m6i.2xlarge       8       32        -              us-east-1     0.38
 Azure        Standard_D8s_v5   8       32        -              eastus        0.38
 IBM          bx2-8x32          8       32        -              us-east       0.38
 GCP          n2-standard-8     8       32        -              us-central1   0.39
 Lambda       gpu_1x_a10        30      200       A10:1          us-east-1     0.60
---------------------------------------------------------------------------------------------------
```


<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] All smoke tests: `pytest tests/test_smoke.py --kubernetes`

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,264 @@
+    def get_credential_file_mounts(self) -> Dict[str, str]:
+        return {}
+        # TODO(romilb): Fix the file mounts optimization
+        #  ('config' here clashes with azure config file)
```
This should have been fixed in #2042

Michaelvll reviewed (commented):
Thanks for adding the k8s feature @romilbhardwaj! This is very exciting! I haven't gone through the codes yet, but tried it out with the `kind` setup locally. The cluster launching and job execution went smoothly!

Left some comments first. Do you think we can have the kind local setup instructions and scripts in a place in the sky module instead of the `tests` folder? I would imagine that would be the first step people would like to try this feature out.

dongreenberg commented:
This is awesome to see!! Two questions - 1) Can you handle a case where a user has access to multiple k8s clusters? and 2) does this assume a sort of exclusive access to the cluster or can the available resources be shifting dynamically?

romilbhardwaj commented:
@dongreenberg - thanks! To answer your questions:

1. Yes, we plan to support multiple k8s clusters soon. If you have a use-case, would be very happy to have you as a design partner for it :)
2. No, we do not assume exclusive access. Resources are requested from Kubernetes on-demand and can have varying availability (e.g., other applications running on the cluster, cluster may be autoscaling). In the event resources are not available, SkyPilot's failover kicks-in and can spill over jobs to other k8s clusters or the cloud.

dongreenberg commented:
You read my mind! We're super interested. Should we chat next week or you think you have a few more weeks to go before there's anything actionable?

romilbhardwaj commented:
Following up on slack :)

romilbhardwaj commented on tests/kubernetes/kind/create_cluster.sh:
```diff
@@ -0,0 +1,11 @@
+# Be sure to have built the latest image before running this script
```
Good idea - refactored into `sky local up`

romilbhardwaj commented on tests/kubernetes/README.md:
````diff
@@ -0,0 +1,43 @@
+
+```bash 
+cd kind
+./create_cluster.sh
````
Ah that was because opening up all the nodeports takes time for some reason.. I've reduced the nodeport range to 30000-30100. Now it starts in ~2min on a gcp node.

romilbhardwaj commented on sky/skylet/providers/kubernetes/utils.py:
```diff
@@ -0,0 +1,72 @@
+import kubernetes
```
Done! Moved into its own adapter now.. still need to verify that import errors don't happen though.

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,264 @@
+
+    SKY_SSH_KEY_SECRET_NAME = f'sky-ssh-{common_utils.get_user_hash()}'
+    _DEFAULT_NUM_VCPUS = 4
+    _DEFAULT_MEMORY_CPU_RATIO = 1
```
After our discussion, changed to 2 CPU by default.

romilbhardwaj commented on tests/kubernetes/README.md:
````diff
@@ -0,0 +1,43 @@
+./build.sh -p
+```
+
+## Running a local development cluster
+You can use (kind)[https://kind.sigs.k8s.io/] to run a local Kubernetes cluster 
+for development. The following script will create a cluster with 1 node and 
+will make NodePort services available on localhost. 
+
+```bash 
+cd kind
+./create_cluster.sh
+```
+
+## Running a GKE cluster
+1. Make sure ports 30000-32767 are open in your node pool VPC's firewall.
+2. Create a GKE cluster with at least 1 node.
+3. Get the kubeconfig for your cluster and place it in `~/.kube/config`:
+```bash
+gcloud container clusters get-credentials <cluster-name> --region <region>
+# Example:
+# gcloud container clusters get-credentials testcluster --region us-central1-c
+```
+4. Verify by running `kubectl get nodes`. You should see your nodes.
+5. You can run SkyPilot tasks now. 
````
Yes! I plan to add this to the docs

romilbhardwaj commented:
Update: added `sky local up` and `sky local down` CLIs to reduce friction in make creating local clusters. Try it out!

Tests (`pytest tests/test_smoke.py --kubernetes`) also appear to be passing. Code needs a little clean up.

Michaelvll commented:
Thanks for the updates @romilbhardwaj! The `sky local up` is very useful for testing the k8s support!
I met the following issue, when the local k8s is supported. 
```
sky launch --gpus V100

    _fill_in_launchable_resources(
  File "/home/gcpuser/skypilot/sky/optimizer.py", line 996, in _fill_in_launchable_resources
    cloud.get_feasible_launchable_resources(resources))
  File "/home/gcpuser/skypilot/sky/clouds/kubernetes.py", line 314, in get_feasible_launchable_resources
    raise NotImplementedError('GPUs are not supported for Kubernetes '
NotImplementedError: GPUs are not supported for Kubernetes clusters yet.
```

Should we include the GPU in the `cloud.CloudImplementationFeatures` enumerator to have failover work correctly?

romilbhardwaj commented:
Thanks for catching that @Michaelvll! I fixed it by simply returning an empty list in `get_feasible_launchable_resources` when gpus are requested (instead of throwing an error):

```
(base) ➜  tmp-workdir SKYPILOT_DEBUG=0 sky launch --gpus V100
I 07-03 09:49:33 optimizer.py:636] == Optimizer ==
I 07-03 09:49:33 optimizer.py:647] Target: minimizing cost
I 07-03 09:49:33 optimizer.py:659] Estimated cost: $2.5 / hour
I 07-03 09:49:33 optimizer.py:659]
I 07-03 09:49:33 optimizer.py:732] Considered resources (1 node):
I 07-03 09:49:33 optimizer.py:781] -----------------------------------------------------------------------------------------------
I 07-03 09:49:33 optimizer.py:781]  CLOUD   INSTANCE           vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN
I 07-03 09:49:33 optimizer.py:781] -----------------------------------------------------------------------------------------------
I 07-03 09:49:33 optimizer.py:781]  IBM     gx2-8x64x1v100     8       64        V100:1         us-east       2.50          ✔
I 07-03 09:49:33 optimizer.py:781]  GCP     n1-highmem-8       8       52        V100:1         us-central1   2.95
I 07-03 09:49:33 optimizer.py:781]  AWS     p3.2xlarge         8       61        V100:1         us-east-1     3.06
I 07-03 09:49:33 optimizer.py:781]  Azure   Standard_NC6s_v3   6       112       V100:1         eastus        3.06
I 07-03 09:49:33 optimizer.py:781] -----------------------------------------------------------------------------------------------
I 07-03 09:49:33 optimizer.py:781]
Launching a new cluster 'sky-197c-romilb'. Proceed? [Y/n]: ^CAborted!
(base) ➜  tmp-workdir SKYPILOT_DEBUG=0 sky launch -c test --gpus V100:1
sky.exceptions.ResourcesMismatchError: Requested resources do not match the existing cluster.
  Requested:	1x <Cloud>({'V100': 1})
  Existing:	1x Kubernetes(2CPU--2GB)
To fix: specify a new cluster name, or down the existing cluster first: sky down test
```

Michaelvll commented on tests/kubernetes/build_image.sh:
```diff
@@ -0,0 +1,45 @@
+fi
+
+echo "Tagging image as skypilot:latest"
+docker tag us-central1-docker.pkg.dev/skypilot-375900/skypilotk8s/skypilot:latest skypilot:latest
```
add a new line for the end of file?

Michaelvll reviewed (commented):
Thanks for adding this exciting feature @romilbhardwaj! I went through all the codes except for some details in the code of the providers, and most of the code looks good to me! Left several comments

romilbhardwaj commented on sky/adaptors/kubernetes.py:
```diff
@@ -0,0 +1,119 @@
+
+def import_package(func):
+
+    @wraps(func)
```
Hmm, I do import the modules at L27 and L28... Do you mean changing the API methods below (e.g.`core_api()`?)

Michaelvll replied:
Oh, I mean we probably need to use `@functools.wraps` instead of `@wraps`.

romilbhardwaj commented on sky/authentication.py:
```diff
@@ -446,3 +447,45 @@ def setup_scp_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
+    except subprocess.CalledProcessError as e:
+        output = e.output.decode('utf-8')
+        suffix = ''
+        if env_options.Options.SHOW_DEBUG_INFO.get():
+            suffix = f' Error message: {output}'
```
Good idea - updated to show the error for non-debugging case as well.

romilbhardwaj commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2466,7 +2538,7 @@ def _provision(
                         handle.cluster_yaml)
                     runners = command_runner.SSHCommandRunner.make_runner_list(
-                        ip_list, **ssh_credentials)
+                        ip_list, **ssh_credentials, port_list=ssh_port_list)
```
That's a good point - I've updated all calls to `make_runner_list`. Do you think we should also move up `port_list` in the method signature to be the first optional arg? i.e., something like:
```
    @staticmethod
    def make_runner_list(
        ip_list: List[str],
        ssh_user: str,
        ssh_private_key: str,
        port_list: Optional[List[int]] = None,
        ssh_control_name: Optional[str] = None,
        ssh_proxy_command: Optional[str] = None,
    ) -> List['SSHCommandRunner']:
```

Or its ok to keep it at the end?

romilbhardwaj commented on sky/__init__.py:
```diff
@@ -49,6 +50,7 @@
     'OptimizeTarget',
     'backends',
     'benchmark',
+    'Kubernetes',
```
Good point, and in addition I've updated the order among clouds to be alphabetical.

romilbhardwaj commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -887,6 +888,34 @@ def _update_blocklist_on_lambda_error(
+                raise RuntimeError('Errors occurred during provisioning; '
+                                   'check logs above.')
+
+        logger.warning(f'Got error(s) in {region.name}:')
```
Current implementation defines only one region `kubernetes`. That will be shown here. Currently we do not support adding more regions (e.g., many kubernetes clusters).. we should discuss that design (i.e., should we add more Kubernetes "clouds" to SkyPilot or add more regions to one Kubernetes "cloud").

romilbhardwaj commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2175,6 +2237,13 @@ def head_ip(self):
+    @property
+    def head_ssh_port(self):
+        external_ssh_ports = self.external_ssh_ports()
+        if external_ssh_ports is not None:
```
ah good catch!

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,353 @@
+                                     region: Optional[str] = None,
+                                     zone: Optional[str] = None) -> float:
+        # Assume zero cost for Kubernetes clusters
+        return 0.0
```
Agreed - added a todo

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,353 @@
+        # TODO(romilb): All accelerators are marked as available for now. In the
+        #  future, we should return false for accelerators that we know are not
+        #  supported by the cluster.
+        return True
```
Ah yeah, good catch - this should return False. Fixed now.

romilbhardwaj commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2175,6 +2237,13 @@ def head_ip(self):
+        return None
+
     def __setstate__(self, state):
         self._version = self._VERSION
```
Done! Incremented schema version to 4 and added handling for `head_ssh_port`

romilbhardwaj commented on sky/skylet/providers/kubernetes/node_provider.py:
```diff
@@ -0,0 +1,371 @@
+        except kubernetes.api_exception():
+            pass
+
+    def terminate_nodes(self, node_ids):
+        for node_id in node_ids:
+            self.terminate_node(node_id)
```
Ah good catch - instead of removing, I have put in a TODO to keep track of this.

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,353 @@
+            memory: Optional[str] = None,
+            disk_tier: Optional[str] = None) -> Optional[str]:
+        del disk_tier  # Unused.
+        # TODO(romilb): Allow fractional CPUs and memory
```
Added!

romilbhardwaj commented on sky/backends/backend_utils.py:
```diff
@@ -937,6 +956,10 @@ def write_cluster_config(
                 'gcp_project_id': gcp_project_id,
 
+                # Kubernetes only:
+                'skypilot_k8s_image': k8s_image,
```
Done - also added a TODO for future PRs to add support for using image specified by user in YAML :)

romilbhardwaj commented on sky/cli.py:
```diff
@@ -4359,6 +4364,99 @@ def _delete_benchmark(benchmark: str) -> None:
+
+@local.command('up', cls=_DocumentedCodeCommand)
+@usage_lib.entrypoint
+def local_up():
```
Thanks for this suggestion - I considered it, but I think maybe lets keep it in `cli.py` for now because it is experimental and adding it to core.py would expose it to python API users. I can leave a TODO to move to core.py when it is no longer experimental. Alternatively, if you think we should have it in core.py in this PR, I can move it there too.

Michaelvll replied:
Sounds good! I think having a TODO for moving is good enough. : )

romilbhardwaj commented on sky/cli.py:
```diff
@@ -4359,6 +4364,99 @@ def _delete_benchmark(benchmark: str) -> None:
+
+@local.command('up', cls=_DocumentedCodeCommand)
+@usage_lib.entrypoint
+def local_up():
```
Yeah, that's a good point. I want to avoid using `launch` because  I want users to associate `launch` with SkyPilot clusters. This CLI should use similar, but not the same, commands, hence I choose `up`.  wdyt?

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,353 @@
+                          f'check if {_CREDENTIAL_PATH} exists.'
+
+    def get_credential_file_mounts(self) -> Dict[str, str]:
+        return {_CREDENTIAL_PATH: _CREDENTIAL_PATH}
```
Ah yeah, that's right - but I think `check_credentials()` should guard against marking k8s as enabled on the spot controller because it includes a reachability check (it tries to run a simple API call on the k8s API server, and if it fails it marks it as not disabled). Do you think that would be enough?

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,353 @@
+    _DEFAULT_NUM_VCPUS = 2
+    _DEFAULT_MEMORY_CPU_RATIO = 1
+    _REPR = 'Kubernetes'
+    _regions: List[clouds.Region] = [clouds.Region('kubernetes')]
```
I agree, the question whether multiple Kubernetes clusters should be represented as different regions or different clouds needs more thought before we add this support in future PRs. This PR aims at supporting only a single Kubernetes cluster, so I added only one "region" to Kubernetes cloud.

romilbhardwaj commented:
Thanks @Michaelvll! I've addressed most of the comments, except for [this](https://github.com/skypilot-org/skypilot/pull/2096#discussion_r1251420820) (which I hope to get to next week).

romilbhardwaj commented on sky/authentication.py:
```diff
@@ -446,3 +446,33 @@ def setup_scp_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
+    config['auth']['ssh_public_key'] = PUBLIC_SSH_KEY_PATH
+
+    file_mounts = config['file_mounts']
+    file_mounts[PUBLIC_SSH_KEY_PATH] = PUBLIC_SSH_KEY_PATH
```
Ahh thanks for the pointer - I've removed public key upload since it is not necessary.

romilbhardwaj commented:
@Michaelvll - thanks for the pointer on public key upload - I've removed it since it was not necessary. Should be ready for another look!

Michaelvll reviewed (commented):
Thanks for the update @romilbhardwaj! This is very exciting! Only some small comments left for the codes. I am now trying it out again. If everything goes well, I think it should be good to go.

romilbhardwaj commented on sky/backends/cloud_vm_ray_backend.py:
```diff
@@ -2315,18 +2313,25 @@ def __setstate__(self, state):
             state['stable_internal_external_ips'] = None
+        if version < 4:
+            # Version 4 adds self.external_ssh_ports for Kubernetes support
+            head_ssh_port = state.pop('head_ssh_port', None)
```
Good catch - fixed now

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,364 @@
+
+    def get_feasible_launchable_resources(self,
+                                          resources: 'resources_lib.Resources'):
+        if resources.use_spot:
```
Since disk tier is a generally a soft-constraint (i.e., job won't fail with a lower disk-tier, would just run slower), I was thinking it's fine to "ignore" disk tier for k8s to enable failover when disk tier is specified. wdyt? (I don't feel too strongly, If we want to be consistent with lambda I can change it here)

romilbhardwaj commented on sky/skylet/providers/kubernetes/utils.py:
```diff
@@ -0,0 +1,63 @@
+        str: Error message if credentials are invalid, None otherwise
+    """
+    try:
+        kubernetes.core_api().list_namespace(_request_timeout=timeout)
```
Ah thanks! Updated now:

```
(base) ➜  ~ sky check
Checking credentials to enable clouds for SkyPilot.
  AWS: enabled
  Azure: enabled
  GCP: enabled
  Lambda: enabled
  IBM: enabled
  SCP: disabled
    Reason: Failed to access SCP with credentials. To configure credentials, see: https://cloud.samsungsds.com/openapiguide
    Generate API key and add the following line to ~/.scp/scp_credential:
      access_key = [YOUR API ACCESS KEY]
      secret_key = [YOUR API SECRET KEY]
      project_id = [YOUR PROJECT ID]
  OCI: disabled
    Reason: `oci` is not installed. Install it with: pip install oci
    For more details, refer to: https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#oracle-cloud-infrastructure-oci
  Kubernetes: disabled
    Reason: `kubernetes` package is not installed. Install it with: pip install kubernetes
  Cloudflare (for R2 object store): enabled
```

romilbhardwaj commented on sky/cli.py:
```diff
@@ -4403,6 +4408,99 @@ def _delete_benchmark(benchmark: str) -> None:
+        path_to_package = os.path.dirname(os.path.dirname(__file__))
+        up_script_path = os.path.join(path_to_package, 'sky/utils/kubernetes',
+                                      'create_cluster.sh')
+        subprocess_utils.run_no_outputs('chmod +x {}'.format(up_script_path))
```
Great point - committed the file in +x mode and removed this chmod line

romilbhardwaj commented on sky/templates/kubernetes-ray.yml.j2:
```diff
@@ -0,0 +1,334 @@
+    pip3 --version > /dev/null 2>&1 || (curl -sSL https://bootstrap.pypa.io/get-pip.py -o get-pip.py && python3 get-pip.py && echo "PATH=$HOME/.local/bin:$PATH" >> ~/.bashrc);
+    (type -a python | grep -q python3) || echo 'alias python=python3' >> ~/.bashrc;
+    (type -a pip | grep -q pip3) || echo 'alias pip=pip3' >> ~/.bashrc;
+    which conda > /dev/null 2>&1 || (wget -nc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh && bash Miniconda3-latest-Linux-x86_64.sh -b && eval "$(~/miniconda3/bin/conda shell.bash hook)" && conda init && conda config --set auto_activate_base true);
```
fixed!

romilbhardwaj commented on tests/kubernetes/build_image.sh:
```diff
@@ -42,4 +42,4 @@ else
 echo "Tagging image as skypilot:latest"
-docker tag us-central1-docker.pkg.dev/skypilot-375900/skypilotk8s/skypilot:latest skypilot:latest
\ No newline at end of file
+docker tag us-central1-docker.pkg.dev/skypilot-375900/skypilotk8s/skypilot:latest skypilot:latest
```
That's a good point this image requires SSH also to be configured. I'm leaving these as separate images for now while we're developing kubernetes support, and I've left a TODO to see if they can be consolidated into the same image.

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,364 @@
+        #  https://kubernetes.io/blog/2022/12/05/forensic-container-checkpointing-alpha/ # pylint: disable=line-too-long
+        clouds.CloudImplementationFeatures.STOP: 'Kubernetes does not '
+                                                 'support stopping VMs.',
+        clouds.CloudImplementationFeatures.AUTOSTOP: 'Kubernetes does not '
```
Ah this is a good point. I've filed #2249 and will create another PR for this since this requires changes (selecting autostop based on cloud being provisioned) which are slightly orthogonal to k8s support.

romilbhardwaj commented:
Thanks @Michaelvll! I've addressed your comments, ready for another look!

- [x] tested: `pytest tests/test_smoke.py --kubernetes -k "not TestStorageWithCredentials"`

Michaelvll reviewed (approved):
Thanks for adding this Kubernetes support @romilbhardwaj! I tried out this PR locally, and it works very well. I think it should be good to go, once we fixed the autodown.
The only problem I met was that:
1. `sky launch --cpus 2 echo hi -i 0 --down` does not correctly terminate the cluster.

romilbhardwaj commented on sky/skylet/providers/kubernetes/utils.py:
```diff
@@ -0,0 +1,63 @@
+                      'Check if your cluster is running and your network ' \
+                      'is stable.'
+    except Exception as e:
+        return False, f'An error occurred: {str(e)}'
```
Ahh that's a good catch. I've added special handling for this case - does this help?
```
(base) ➜  ~ SKYPILOT_DEBUG=0 sky check
Checking credentials to enable clouds for SkyPilot.
  AWS: enabled
  Azure: enabled
  GCP: enabled
  Lambda: enabled
  IBM: enabled
  SCP: disabled
    Reason: Failed to access SCP with credentials. To configure credentials, see: https://cloud.samsungsds.com/openapiguide
    Generate API key and add the following line to ~/.scp/scp_credential:
      access_key = [YOUR API ACCESS KEY]
      secret_key = [YOUR API SECRET KEY]
      project_id = [YOUR PROJECT ID]
  OCI: disabled
    Reason: `oci` is not installed. Install it with: pip install oci
    For more details, refer to: https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#oracle-cloud-infrastructure-oci
  Kubernetes: disabled
    Reason: Failed to load Kubernetes configuration. Kubeconfig does not contain any valid context(s).
    If you were running a local Kubernetes cluster, run `sky local up` to start the cluster.
  Cloudflare (for R2 object store): enabled

SkyPilot will use only the enabled clouds to run tasks. To change this, configure cloud credentials, and run sky check.
If any problems remain, please file an issue at https://github.com/skypilot-org/skypilot/issues/new
```

romilbhardwaj commented on sky/clouds/kubernetes.py:
```diff
@@ -0,0 +1,364 @@
+    SKY_SSH_KEY_SECRET_NAME = f'sky-ssh-{common_utils.get_user_hash()}'
+
+    # TODO(romilb): Make the timeout configurable.
+    TIMEOUT = 60  # Timeout for resource provisioning
```
TL;DR - good catch, I've reduced it to 10s (tested with `sky local up` and GKE non-autoscaling cluster).

----
The reason we have this timeout is to 1) give time to Kubernetes to schedule the pods and 2) support autoscaling in the future. Since Kubernetes keeps pods indefinitely pending till resources are available AND we don't have visibility into any underlying autoscaling, we add a timeout to give the underlying autoscaler enough time to add resources. In the future, this timeout will be configurable by the user depending on the following:

* For non-autoscaling clusters - this timeout should be close to zero, since if the pod doesn't get scheduled immediately, we know for sure there are not resources currently available.

* For autoscaling clusters - this timeout would need to be set to roughly the amount of time it takes to add a new node.

Previously, this timeout also included the time taken to pull the image so I had set it conservatively to 60s. We now break this timeout check loop the moment a pod starts pulling an image (since image pull happens only after pod has been scheduled). Thus, this can safely be reduced to only account for scheduling time, and I found that 10s is a safe timeout to use.

romilbhardwaj commented:
Thanks @Michaelvll! Good catch with the autodown. I have fixed it (and `sky status --referesh`) now and also added a smoke test `test_kubernetes_autodown`. I've also addressed other comments.

Tested:
- [x] `pytest tests/test_smoke.py --kubernetes`
- [ ] all smoke tests `pytest tests/test_smoke.py`

Michaelvll commented on sky/skylet/providers/kubernetes/utils.py:
```diff
@@ -47,22 +50,41 @@ def check_credentials(timeout: int = 3) -> Tuple[bool, Optional[str]]:
+    except ValueError as e:
+        return False, str(e)
     except Exception as e:
         return False, f'An error occurred: {str(e)}'
```
nit: A better way to print the exceptions will be `common_utils.format_exception(e)` so the type of the exception will b printed, making it easier to debug.

Michaelvll commented on sky/templates/kubernetes-ray.yml.j2:
```diff
@@ -46,10 +42,12 @@ provider:
             labels:
                 parent: skypilot
             name: autoscaler
+        # TODO(romilb): This is a very permissive role - gives all access in the
+        #  namespace. We should restrict this.
```
Should we mention that this is for the `autodown` to work, so in the future we know how to test the restricted permissions?

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -358,7 +372,19 @@ def accelerator_in_region_or_zone(self,
-        #  Assuming single node cluster.
         del tag_filters, region, zone, kwargs  # Unused.
-        return [status_lib.ClusterStatus.UP]
+        namespace = kubernetes_utils.get_current_kube_config_context_namespace()
+
+        # Get all the pods with the label skypilot-cluster: <cluster_name>
+        pods = kubernetes.core_api().list_namespaced_pod(
+            namespace, label_selector=f'skypilot-cluster={name}').items
```
We should catch the exceptions here to raise only `exceptions.ClusterStatusFetchingError`, as our callers assumes the exceptions to be `exceptions.ClusterStatusFetchingError` for this `query_status` method.
https://github.com/skypilot-org/skypilot/blob/f5526d43b49f5e8bf184f73c58193d9d203cdc1e/sky/backends/backend_utils.py#L2350-L2365
https://github.com/skypilot-org/skypilot/blob/f5526d43b49f5e8bf184f73c58193d9d203cdc1e/sky/backends/backend_utils.py#L2183-L2206

To reproduce:
1. `sky launch --cpus 2+` (on k8s)
2. `sky local down`
3. `sky status -r` fails and my other clouds' cluster fail to show up.
```
$ sky status -r
Clusters
ValueError: Failed to load Kubernetes configuration. Kubeconfig does not contain any valid context(s).
    If you were running a local Kubernetes cluster, run `sky local up` to start the cluster.
```

Michaelvll replied:
Another question is: if the user switches their `~/.kube/config`, the existing cluster on the original k8s will be leaked if `sky status -r` is called?

That might be fixed by implementing the `get_current_user_identity` in the future, so that the cluster will be marked with the current kubernetes cluster, and will not be operated when it does not match.

romilbhardwaj replied:
Ah that's a great catch - I've fixed `sky status -r` now to raise `ClusterStatusFetchingError` when the kubernetes API call fails:
```
# With a empty kubeconfig after sky local down
(base) ➜  ~ sky status -r
Clusters
W 07-25 18:42:14 backend_utils.py:2441] Failed to refresh status for 1 cluster:
W 07-25 18:42:14 backend_utils.py:2444]   test: Failed to query Kubernetes cluster 'test' status: builtins.ValueError: Failed to load Kubernetes configuration. Kubeconfig does not contain any valid context(s).
W 07-25 18:42:14 backend_utils.py:2444]     If you were running a local Kubernetes cluster, run `sky local up` to start the cluster.
NAME  LAUNCHED     RESOURCES                          STATUS  AUTOSTOP  COMMAND
test  36 secs ago  1x Kubernetes(2CPU--2GB, cpus=2+)  UP      -         sky launch --cpus 2+ -c t...

Managed spot jobs
No in progress jobs. (See: sky spot -h)

# With an valid kubeconfig but bad IP address/cluster not reachable
(base) ➜  ~ sky status -r
Clusters
W 07-25 18:43:59 backend_utils.py:2441] Failed to refresh status for 1 cluster:
W 07-25 18:43:59 backend_utils.py:2444]   test: Failed to query cluster 'test' status. Network error - check if the Kubernetes cluster in context gke_skypilot-375900_us-central1-c_testclus2 is up and accessible.
NAME  LAUNCHED    RESOURCES                          STATUS  AUTOSTOP  COMMAND
test  2 mins ago  1x Kubernetes(2CPU--2GB, cpus=2+)  UP      -         sky launch --cpus 2+ -c t...

Managed spot jobs
No in progress jobs. (See: sky spot -h)
```

> Another question is: if the user switches their ~/.kube/config, the existing cluster on the original k8s will be leaked if sky status -r is called?

This is a good point, currently the cluster will indeed be leaked. I've filed #2300 to track and fix this.

Michaelvll reviewed (approved):
Thanks for the quick fix of the autodown @romilbhardwaj! LGTM with a problem of cluster refresh exceptions.

Michaelvll commented on sky/authentication.py:
```diff
@@ -372,3 +372,33 @@ def setup_scp_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
+        output = e.output.decode('utf-8')
+        suffix = f'\nError message: {output}'
+        if 'already exists' in output:
+            logger.warning(
+                f'Key {key_label} already exists in the cluster, using it...')
```
Do we need to print this out? It seems a bit confusing

romilbhardwaj replied:
Ahh good point. I used it for debugging. Changed the `warning` to `debug`.

Michaelvll reviewed (commented):
Just tried this on the GKE cluster with and without autoscaling. It works smoothly, except that for the one with autoscaling, the first `sky launch` that trigger the autoscaling will timeout, because the GKE takes longer to set up the cluster?

romilbhardwaj commented:
I've removed ingress creation since it was no longer required, and confirmed it now works with older client library (`kubernetes==17.17.0`) forced by `PyYAML<=5.3.1`. 

Tested:
- [x] `pytest tests/test_smoke.py --kubernetes`

> It works smoothly, except that for the one with autoscaling, the first sky launch that trigger the autoscaling will timeout, because the GKE takes longer to set up the cluster?

That's right, autoscaling takes time and that's why we want to make the provisioning [timeout configurable](https://github.com/skypilot-org/skypilot/blob/4fb1d9427b9d7968299d970945e020037dc0c513/sky/clouds/kubernetes.py#L133-L141) in the near future so users can set it according to their cluster. In the medium-long term, we should look in to interfacing with the underlying autoscaler(s) to know when it is scaling up, and automatically change the timeout as needed.

Michaelvll commented on sky/clouds/kubernetes.py:
```diff
@@ -384,13 +384,13 @@ def query_status(cls, name: str, tag_filters: Dict[str, str],
                 label_selector=f'skypilot-cluster={name}',
                 _request_timeout=kubernetes.API_TIMEOUT).items
-        except kubernetes.max_retry_error():
+        except kubernetes.max_retry_error() as e:
```
nit: Seems `e` is not used here?

Michaelvll reviewed (approved):
Thanks for the great effort @romilbhardwaj! LGTM. Very excited to see the k8s support to be merged.

romilbhardwaj commented:
Thanks for the reviews @Michaelvll! All smoke tests pass, merging now.
